src/main/java/org/opensearch/knn/jni/FaissService.java

```diff
@@ -28,16 +36,78 @@
+    private static Logger logger = LogManager.getLogger(FaissService.class);
 
     static {
         AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
-            System.loadLibrary(KNNConstants.FAISS_JNI_LIBRARY_NAME);
+
+            // Even if the underlying system supports AVX2, users can override and disable it by using the
+            // 'knn.faiss.avx2.disabled' setting by setting it to true in the opensearch.yml configuration
+            if (!isFaissAVX2Disabled() && isAVX2SupportedBySystem()) {
+                System.loadLibrary(KNNConstants.FAISS_AVX2_JNI_LIBRARY_NAME);
+            } else {
+                System.loadLibrary(KNNConstants.FAISS_JNI_LIBRARY_NAME);
+            }
+
             initLibrary();
             KNNEngine.FAISS.setInitialized(true);
             return null;
         });
     }
 
+    /**
+     * Verify if the underlying system supports AVX2 SIMD Optimization or not
+     * 1. If the architecture is aarch64 return false.
+     * 2. If the operating system is windows return false.
+     * 3. If the operating system is macOS, use oshi-core library to verify if the cpu flags
+     *    contains 'avx2' and return true if it exists else false.
+     * 4. If the operating system is linux, read the '/proc/cpuinfo' file path and verify if
+     *    the flags contains 'avx2' and return true if it exists else false.
+     */
+    private static boolean isAVX2SupportedBySystem() {
+        if ((System.getProperty("os.arch").toLowerCase()).contains("aarch")
+            || (System.getProperty("os.name").toLowerCase()).contains("windows")) {
+            return false;
+        }
+
+        if ((System.getProperty("os.name").toLowerCase()).contains("mac")) {
+            try {
+                return AccessController.doPrivileged((PrivilegedExceptionAction<Boolean>) () -> {
+                    String flags = SysctlUtil.sysctl("machdep.cpu.leaf7_features", "empty");
+                    if ((flags.toLowerCase()).contains("avx2")) {
+                        return true;
+                    }
+                    return false;
+                });
+            } catch (Exception e) {
+                logger.error("[KNN] Error fetching cpu flags info. [{}]", e.getMessage());
+                e.printStackTrace();
+            }
+
+        } else if ((System.getProperty("os.name").toLowerCase()).contains("linux")) {
+            String fileName = "/proc/cpuinfo";
+            try {
+                return AccessController.doPrivileged((PrivilegedExceptionAction<Boolean>) () -> {
+                    String cpuFlags = Files.lines(Paths.get(fileName))
+                        .filter(s -> s.startsWith("flags"))
+                        .filter(s -> s.contains("avx2"))
+                        .findFirst()
+                        .orElse("");
+
+                    if ((cpuFlags.toLowerCase()).contains("avx2")) {
+                        return true;
+                    }
+                    return false;
```

Comment:
This could be improved. You're already filtering for `avx2` (before doing lower case) so this seems redundant.  And an "if (boolean) { return true; } return false;" can just be "return (boolean)".

Suggest changing `findFirst()` to `findAny()` and then using `map(String::isEmpty)` to produce an optional boolean; then the orElse can be true or false.